net/httpclient_example_test.go

```diff
@@ -325,3 +326,74 @@ func ExampleClient_hostSecret() {
 		time.Sleep(1 * time.Second)
 	}
 }
+
+func ExampleClient_withBeforeSendHook() {
+	mockTracer := mocktracer.New()
+	peerService := "my-peer-service"
+	cli := net.NewClient(net.Options{
+		Tracer:                     &customTracer{mockTracer},
+		OpentracingComponentTag:    "testclient",
+		OpentracingSpanName:        "clientSpan",
+		BearerTokenRefreshInterval: 10 * time.Second,
+		BearerTokenFile:            "/tmp/foo.token",
+		IdleConnTimeout:            2 * time.Second,
+		BeforeSend: func(req *http.Request) {
+			req.Header.Set("X-Foo", "qux")
+			if span := opentracing.SpanFromContext(req.Context()); span != nil {
+				span.SetTag(string(ext.PeerService), peerService)
+			}
+		},
+	})
+	defer cli.Close()
+
+	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		fmt.Printf("X-Foo: %s\n", r.Header.Get("X-Foo"))
+		w.WriteHeader(http.StatusOK)
+	}))
+	defer srv.Close()
+
+	cli.Get("http://" + srv.Listener.Addr().String() + "/")
+	fmt.Printf("request tag %q set to %q", string(ext.PeerService), mockTracer.FinishedSpans()[0].Tags()[string(ext.PeerService)])
+
+	// Output:
+	// X-Foo: qux
+	// request tag "peer.service" set to "my-peer-service"
+}
+
+func ExampleClient_withAfterResponseHook() {
+	mockTracer := mocktracer.New()
+	cli := net.NewClient(net.Options{
+		Tracer:                     &customTracer{mockTracer},
+		OpentracingComponentTag:    "testclient",
+		OpentracingSpanName:        "clientSpan",
+		BearerTokenRefreshInterval: 10 * time.Second,
+		BearerTokenFile:            "/tmp/foo.token",
+		IdleConnTimeout:            2 * time.Second,
+		AfterResponse: func(rsp *http.Response, err error) {
+			if span := opentracing.SpanFromContext(rsp.Request.Context()); span != nil {
+				span.SetTag("status.code", rsp.StatusCode)
+				if err != nil {
+					span.SetTag("error", err.Error())
+				}
+			}
+			rsp.StatusCode = 255
+		},
+	})
+	defer cli.Close()
+
+	srv := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusOK)
+	}))
+	defer srv.Close()
+
+	rsp, err := cli.Get("http://" + srv.Listener.Addr().String() + "/")
+	if err != nil {
+		log.Fatalf("Failed to get: %v", err)
+	}
+	fmt.Printf("response code: %d\n", rsp.StatusCode)
+	fmt.Printf("span status.code: %d", mockTracer.FinishedSpans()[0].Tags()["status.code"])
```

Comment:
Client may return before span is available in finished spans.
To mitigate you'd need to wait for it, see example https://github.com/zalando/skipper/commit/34b188cab70fe01bed88b6a9a16db7a34559a70d#diff-a12e7212c817eb6b6e6168aaeeb1e1faefd34de890a929adc152505bf34bf097R268-R269

Reply:
Because we have no testing.T, I will create a loop that waits for some time similar to the linked code.

Reply:
I added some debug logs to figure out what's going on, not sure how this can actually happen:
```
% go test ./net -run ExampleClient_withBeforeSendHook
2024/07/03 17:28:31 injectSpan
time="2024-07-03T17:28:31+02:00" level=info msg="BeforeSend: no span found"
time="2024-07-03T17:28:31+02:00" level=info msg="found 1 spans"
--- FAIL: ExampleClient_withBeforeSendHook (0.00s)
got:
X-Foo: qux
request tag "peer.service" set to %!q(<nil>)
want:
X-Foo: qux
request tag "peer.service" set to "my-peer-service"
FAIL
FAIL    github.com/zalando/skipper/net  0.099s
FAIL
```